ntoskrnl/io/iomgr/deviface.c

```diff
@@ -26,192 +27,498 @@ C_ASSERT(sizeof(L"{01234567-89ab-cdef-0123-456789abcdef}") == GUID_STRING_BYTES
+    ReferenceStringReal.Length = LinkNameNoPrefix.Length - ReferenceStringOffset;
+    ReferenceStringReal.MaximumLength = ReferenceStringReal.Length;
+
+    /* Get GUID string (device class GUID in {} brackets) */
+    GuidStringReal.Buffer = LinkNameNoPrefix.Buffer + (ReferenceStringOffset - GUID_STRING_BYTES) / sizeof(WCHAR);
+    GuidStringReal.Length = GUID_STRING_BYTES;
+    GuidStringReal.MaximumLength = GuidStringReal.Length;
+
+    /* Create a temporary device string (looks like ACPI#PNP0501#1#) */
+    DeviceString.Buffer = LinkNameNoPrefix.Buffer;
+    DeviceString.Length = LinkNameNoPrefix.Length - ReferenceStringReal.Length;
+    DeviceString.MaximumLength = DeviceString.Length;
+
+    DPRINT("Device string is '%wZ'\n", &DeviceString);
+
+    if (MungedString)
     {
-        Status = STATUS_INSUFFICIENT_RESOURCES;
-        goto cleanup;
+        /* Allocate a munged path string */
+        MungedStringReal.Length = 0;
+        MungedStringReal.MaximumLength = MungedPrefix.Length + DeviceString.Length + sizeof(UNICODE_NULL);
+        MungedStringReal.Buffer = ExAllocatePoolWithTag(PagedPool, MungedStringReal.MaximumLength, TAG_IO);
+        if (!MungedStringReal.Buffer)
+        {
+            DPRINT1("ExAllocatePoolWithTag() failed\n");
+            return STATUS_INSUFFICIENT_RESOURCES;
+        }
+
+        /* Fill it by munged prefix and our previously created device string */
+        Status = RtlAppendUnicodeStringToString(&MungedStringReal, &MungedPrefix);
+        NT_ASSERT(NT_SUCCESS(Status));
+
+        Status = RtlAppendUnicodeStringToString(&MungedStringReal, &DeviceString);
+        NT_ASSERT(NT_SUCCESS(Status));
+
+        DPRINT("Munged string is '%wZ'\n", &MungedStringReal);
     }
 
-    RtlAppendUnicodeStringToString(&SubKeyName,
-                                   SymbolicLinkName);
+    DPRINT("GUID string is '%wZ'\n", &GuidStringReal);
+    DPRINT("Reference string is '%wZ'\n", &ReferenceStringReal);
+
+    /* Store received parts if the parameters are not null */
+    if (PrefixString)
+    {
+        PrefixString->Buffer = SymbolicLinkName->Buffer;
+        PrefixString->Length = KernelModePrefix.Length; // Same as UserModePrefix.Length
+        PrefixString->MaximumLength = PrefixString->Length;
+
+        DPRINT("Prefix string is '%wZ'\n", PrefixString);
+    }
 
-    SubKeyName.Buffer[SubKeyName.Length / sizeof(WCHAR)] = UNICODE_NULL;
+    if (MungedString)
+        *MungedString = MungedStringReal;
 
-    SubKeyName.Buffer[0] = L'#';
-    SubKeyName.Buffer[1] = L'#';
-    SubKeyName.Buffer[2] = L'?';
-    SubKeyName.Buffer[3] = L'#';
+    if (GuidString)
+        *GuidString = GuidStringReal;
 
-    ReferenceString.Buffer = wcsrchr(SubKeyName.Buffer, '\\');
-    if (ReferenceString.Buffer != NULL)
+    if (ReferenceString)
+        *ReferenceString = ReferenceStringReal;
+
+    if (ReferenceStringPresent)
+        *ReferenceStringPresent = ReferenceStringReal.Length > sizeof(WCHAR);
```

Comment:
So if the reference string has exactly one WCHAR character (so its `Length == sizeof(WCHAR)`), the `*ReferenceStringPresent` would still be FALSE?

Reply:
Yep. It's logically correct that the only backslash `\` without anything else means reference string is not existing, since nothing comes after that. If it has a length of two or more characters, the 2nd and the following space entries should contain another characters and hence everything after the first backslash is considered as reference string. But even if ref string is missing, the 1st character should be initialized anyway for the registry-related stuff (creating/opening `#` registry key) we already discussed above.